[GTK][WPE] Implement RunLoopObserver for glib ports

nikolaszimmermann opened the pull request:
#### 8615019f48b156e271af8f0ae345697c14ed99d7
<pre>
[GTK][WPE] Implement RunLoopObserver for glib ports
<a href="https://bugs.webkit.org/show_bug.cgi?id=300022">https://bugs.webkit.org/show_bug.cgi?id=300022</a>

Reviewed by Carlos Garcia Campos.

Implement RunLoopObserver for GTK/WPE - ports using the GLib event loop.
This has been available on Apple platforms since a decade, and was used to
drive e.g. the rendering prior to GPUProcess. OpportunisticTaskScheduler
functionality (executing GC in &apos;good moments&apos;, not blocking rendering
etc.) depends on RunLoopObserver, which we failed to provide so far.

Add new API tests covering the ActiviyObserver/RunLoopObserver
implementation for the glib ports.

Tests: Tools/TestWebKitAPI/Tests/WTF/glib/ActivityObserver.cpp
       Tools/TestWebKitAPI/Tests/WebCore/glib/RunLoopObserver.cpp

* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/glib/ActivityObserver.h: Added.
(WTF::ActivityObserver::create):
(WTF::ActivityObserver::order const):
(WTF::ActivityObserver::activities const):
(WTF::ActivityObserver::notify const):
(WTF::ActivityObserver::ActivityObserver):
* Source/WTF/wtf/glib/RunLoopGLib.cpp:
(WTF::RunLoop::RunLoop):
(WTF::RunLoop::~RunLoop):
(WTF::RunLoop::runGLibMainLoopIteration):
(WTF::RunLoop::runGLibMainLoop):
(WTF::RunLoop::run):
(WTF::RunLoop::stop):
(WTF::RunLoop::cycle):
(WTF::RunLoop::observeEvent):
(WTF::RunLoop::observeActivity):
(WTF::RunLoop::unobserveActivity):
(WTF::RunLoop::notifyActivity):
(WTF::RunLoop::notifyEvent):
(WTF::RunLoop::observe): Deleted.
(WTF::RunLoop::notify): Deleted.
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp:
(WebCore::PageTimelineAgent::internalStart):
* Source/WebCore/inspector/agents/page/PageTimelineAgent.h:
* Source/WebCore/platform/RunLoopObserver.cpp:
(WebCore::RunLoopObserver::runLoopObserverFired):
* Source/WebCore/platform/RunLoopObserver.h:
(WebCore::RunLoopObserver::RunLoopObserver):
* Source/WebCore/platform/glib/RunLoopObserverGLib.cpp: Added.
(WebCore::RunLoopObserver::schedule):
(WebCore::RunLoopObserver::invalidate):
(WebCore::RunLoopObserver::isScheduled const):
* Tools/TestWebKitAPI/PlatformGTK.cmake:
* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WTF/glib/ActivityObserver.cpp: Added.
(TestWebKitAPI::dispatchChecker):
(TestWebKitAPI::runTestWhileRunLoopIsActive):
(TestWebKitAPI::TEST(WTF_ActivityObserver, Create)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, MatchingActivity)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, NonMatchingActivity)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, MultipleActivities)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, Ordering)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, SameOrder)):
* Tools/TestWebKitAPI/Tests/WebCore/glib/RunLoopObserver.cpp: Added.
(TestWebKitAPI::setUpTest):
(TestWebKitAPI::TEST(RunLoopObserver, Schedule)):
(TestWebKitAPI::TEST(RunLoopObserver, Invalidate)):
(TestWebKitAPI::TEST(RunLoopObserver, MultipleSchedule)):
(TestWebKitAPI::TEST(RunLoopObserver, MultipleInvalidate)):
(TestWebKitAPI::TEST(RunLoopObserver, Destruction)):
(TestWebKitAPI::TEST(RunLoopObserver, Repeating)):
(TestWebKitAPI::TEST(RunLoopObserver, OneShot)):
(TestWebKitAPI::TEST(RunLoopObserver, DefaultActivities)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityEntry)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityExit)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityBeforeWaiting)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityAfterWaiting)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityCombination)):
(TestWebKitAPI::TEST(RunLoopObserver, RemovesSelfDuringCallback)):
(TestWebKitAPI::TEST(RunLoopObserver, AddsNewObserverDuringCallback)):
(TestWebKitAPI::TEST(RunLoopObserver, AcrossMultipleIterations)):
(TestWebKitAPI::TEST(RunLoopObserver, WellKnownOrderValues)):
(TestWebKitAPI::TEST(RunLoopObserver, DifferentWellKnownOrderValues)):

Canonical link: <a href="https://commits.webkit.org/301378@main">https://commits.webkit.org/301378@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a1069df357f7f03477e555757811de2ff1abb5d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125779 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36192 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132647 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77657 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127650 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54000 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95819 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63937 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128727 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36875 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112475 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76311 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35777 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30659 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76120 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117860 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106654 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135329 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124286 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52566 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40315 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104288 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53014 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104016 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26488 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49380 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27698 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49871 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52461 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58270 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/157298 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51809 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39378 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55158 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53503 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
The FIXME was there on purpose, I will remove this code, was just too lazy -- it's clear to me that importing the GMainLoop struct is not the solution :-)

nikolaszimmermann commented:
Thanks @mcatanzaro for the review, it's still in WIP state until I removed the hack of importing GMainLoop internals.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c81c071de2c7db21b9d5df241b1de32754453f0b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51678

nikolaszimmermann commented:
@mcatanzaro your comments helped a lot, I hope the new variant is nicer. Please double-check, the code is yet untested, still waiting for the compilation -- touching RunLoop.h is evil...

carlosgcampos commented on Source/WebCore/platform/RunLoopObserver.h:
```diff
@@ -31,6 +31,10 @@
 #include <wtf/RetainPtr.h>
 #include <wtf/TZoneMalloc.h>
 
+#if USE(GLIB)
+#include <wtf/RunLoop.h>
+#endif
```
This should be included for al ports now

nikolaszimmermann replied:
Unified builds... you're right, fixed.

carlosgcampos commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -141,18 +146,43 @@ RunLoop::CycleResult RunLoop::cycle(RunLoopMode)
+
+void RunLoop::notifyActivity(Activity activity)
+{
+    if (m_activityObservers.isEmpty())
+        return;
+
+    for (const auto& observer : m_activityObservers) {
```
for loop will do nothing if vector is empty so we don't need the early return, I think.

carlosgcampos commented on Source/WTF/wtf/RunLoop.h:
```diff
@@ -114,11 +115,62 @@ class WTF_CAPABILITY("is current") RunLoop final : public GuaranteedSerialFuncti
+    WTF_EXPORT_PRIVATE void observeEvent(const EventObserver&);
+
+    // Activity observers (used to implement WebCore::RunLoopObserver)
+    class ActivityObserver {
```
Can this be moved to its own file? I think it's too much glib specific code for a cross-platform header.

nikolaszimmermann replied:
Moving to glib/ActivityObserver.h.

carlosgcampos commented on Source/WTF/wtf/RunLoop.h:
```diff
@@ -114,11 +115,62 @@ class WTF_CAPABILITY("is current") RunLoop final : public GuaranteedSerialFuncti
+    class ActivityObserver {
+        WTF_DEPRECATED_MAKE_FAST_ALLOCATED(ActivityObserver);
+    public:
+        using Callback = std::function<void()>;
```
Why not just using WTF::Function?

carlosgcampos reviewed (commented):
I think we should add api tests for this.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/61e14420f3e9a94af880e3bdb808e610b0e4b16e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51678

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +82,57 @@ RunLoop::RunLoop()
+    std::array<GPollFD, 128> fds;
+    while (!m_shouldStop) {
+        gint maxPriority = 0;
+        auto hasReadySources = g_main_context_prepare(mainContext, &maxPriority);
```
Oh, that was taken care before by the g_main_context_push_thread_default, no? Michael suggested to drop this IIRC

carlosgcampos replied:
Yes, but Michael was right, we don't need to push/pop we need only the acquire/release

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +82,57 @@ RunLoop::RunLoop()
+        auto nfds = g_main_context_query(mainContext, maxPriority, &timeout, fds.data(), fds.max_size());
+
+        notifyActivity(Activity::BeforeWaiting);
+        if (!hasReadySources) {
```
You mean that `g_main_context_iterate_unlocked` ignores the return value?

carlosgcampos replied:
Yes, the return value of prepare, as well as clutter run loop integration with CFRunLoop

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +82,57 @@ RunLoop::RunLoop()
+        auto hasReadySources = g_main_context_prepare(mainContext, &maxPriority);
+
+        gint timeout = 0;
+        auto nfds = g_main_context_query(mainContext, maxPriority, &timeout, fds.data(), fds.max_size());
```
Using a Vector now with pre-defined inline capacity of 32.

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -141,18 +146,43 @@ RunLoop::CycleResult RunLoop::cycle(RunLoopMode)
+
+void RunLoop::observeActivity(const RunLoop::ActivityObserver& observer)
+{
+    ASSERT(!m_activityObservers.contains(observer));
+    m_activityObservers.append(observer);
```
Done.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/40adb1f45ba90728a2f21e5f54e8caf6a0d3c6d8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51678

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/954dcb61601ff330926af0b70cadcab869f2462f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51678

mcatanzaro commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +84,65 @@ RunLoop::RunLoop()
+
+void RunLoop::runGLibMainLoop(GMainContext* mainContext)
+{
+    g_main_context_ref(mainContext);
```
You can use GRefPtr here.

nikolaszimmermann replied:
m_mainContext is already a GRefPtr, I think the manual ref/derefing is superfluous. Will double check, too.

mcatanzaro commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +84,65 @@ RunLoop::RunLoop()
     }
+
+    notifyActivity(Activity::Exit);
+    g_main_context_unref(mainContext);
+    g_main_context_pop_thread_default(mainContext);
```
This is the wrong order. If it's OK for the unref to not be last, then the unref is not needed. :) (And it's needed.) Another benefit of GRefPtr.

mcatanzaro commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +84,65 @@ RunLoop::RunLoop()
+        if (result < 0) {
+            LOG_ERROR("RunLoop::runGLibMainLoop() - polling failed, exiting.");
+            break;
+        }
```
Looking at `g_main_context_poll_unlocked()` I think you're missing a couple points. First, use `g_main_context_get_poll_func()` rather than assuming `g_poll()`; this could matter if it's the UI process, where we don't control the default main context.

I'm not sure why you switched from acquire/release (which should be sufficient) back to push/pop thread default, though. I assume you have a reason for it; why? This should be safe and harmless, but I would expect code that creates the GMainContext to handle that if needed.

Then regarding the error handling:

 * You can use wtf/SafeStrerror.h to format `errno` and include it in the error message.
 * Except, don't print the error message at all if `errno == EINTR` since that's expected.
 * Normally `EINTR` would indicate the syscall should be retried, but in this case we can safely just ignore it.
 * Is it OK to notify `Activity::BeforeWaiting` and then skip `Activity::AfterWaiting` in the error case?
 * I was also about to complain about the `break` here causing you to skip the `g_main_context_release()`, which was in the wrong place. I see you noticed this problem and pushed a fixed version even as I was typing this.
 * But even so, I don't think the break is correct. If there is an error from `poll()`, then we should probably just keep looping, right? Even in the unlikely event it loops forever, that's better than stopping, which is never going to be correct.

nikolaszimmermann replied:
Still need to tackle this.

mcatanzaro commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +84,65 @@ RunLoop::RunLoop()
+            break;
+        }
+
+        notifyActivity(Activity::AfterWaiting);
```
[Documentation says:](https://developer.apple.com/documentation/corefoundation/cfrunloopactivity)

> Inside the event processing loop after the run loop wakes up, but before processing the event that woke it up. This activity occurs only if the run loop did in fact go to sleep during the current loop.

So, same as BeforeWaiting: I think that means we should notify it only if `timeout > 0`.

nikolaszimmermann replied:
Still need to tackle this.

nikolaszimmermann replied:
Doesn't apply, see next comment.

mcatanzaro commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +84,65 @@ RunLoop::RunLoop()
-        if (!g_main_loop_is_running(m_mainLoops[i].get()))
-            continue;
-        g_main_loop_quit(m_mainLoops[i].get());
+        notifyActivity(Activity::BeforeWaiting);
```
[Documentation says:](https://developer.apple.com/documentation/corefoundation/cfrunloopactivity/beforewaiting)

> Inside the event processing loop before the run loop sleeps, waiting for a source or timer to fire. This activity does not occur if [CFRunLoopRunInMode(_:_:_:)](https://developer.apple.com/documentation/corefoundation/cfrunloopruninmode(_:_:_:)) is called with a timeout of 0 seconds. It also does not occur in a particular iteration of the event processing loop if a version 0 source fires.

I think that means we should notify it only if `timeout > 0`.

nikolaszimmermann replied:
Still need to tackle this.

nikolaszimmermann replied:
I'm not sure about this anymore:

```
This activity does not occur if [CFRunLoopRunInMode(_:_:_:)](https://developer.apple.com/documentation/corefoundation/cfrunloopruninmode(_:_:_:)) is called with a timeout of 0 seconds.
```

it specifically talks about `CFRunLoopRunInMode` called with a timeout of 0.


The CF RunLoop implementation never specifies a timeout of 0. It either calls `CFRunLoopRunInMode` with a `0.05` timeout in `cycle()`, or the `CFRunLoopRun` variant that spins the loop indefinately.

```
RunLoop::CycleResult RunLoop::cycle(RunLoopMode mode)
{
    CFTimeInterval timeInterval = 0.05;
    CFRunLoopRunInMode(mode, timeInterval, true);
    return CycleResult::Continue;
}

void RunLoop::run()
{
    AutodrainedPool pool;
    CFRunLoopRun();
}
```

mcatanzaro commented on Source/WTF/wtf/RunLoop.h:
```diff
@@ -114,11 +118,27 @@ class WTF_CAPABILITY("is current") RunLoop final : public GuaranteedSerialFuncti
+    };
+
+    using EventObserver = Observer<void(Event, const String&)>;
+    WTF_EXPORT_PRIVATE void observeEvent(const EventObserver&);
+
+    WTF_EXPORT_PRIVATE void observeActivity(const RefPtr<ActivityObserver>&);
+    WTF_EXPORT_PRIVATE void unobserveActivity(const RefPtr<ActivityObserver>&);
```
I see that scheduling/unscheduling events and stopping the RunLoop is all thread-safe. Probably the observers should be too? But it is not, because `m_activityObservers` and `m_eventObservers` is not locked.

mcatanzaro reviewed (changes requested):
If this was anything other than RunLoop, I would click the Approve button now and trust you to follow up on my feedback. But: it is RunLoop, so I'm going to be strict. :)

nikolaszimmermann commented:
There are six new API test failures -- all related to callbacks dispatched in main threads, where the expectation from the tests were that they are dispatched in worker threads (WorkQueue related) -- this seems to fix the issue:

```
diff --git a/Source/WTF/wtf/glib/RunLoopGLib.cpp b/Source/WTF/wtf/glib/RunLoopGLib.cpp
index ea380e085198..3b12f5775a5b 100644
--- a/Source/WTF/wtf/glib/RunLoopGLib.cpp
+++ b/Source/WTF/wtf/glib/RunLoopGLib.cpp
@@ -90,13 +90,11 @@ RunLoop::~RunLoop()
 void RunLoop::runGLibMainLoop(GMainContext* mainContext)
 {
     g_main_context_ref(mainContext);
+    g_main_context_push_thread_default(mainContext);
     notifyActivity(Activity::Entry);
 
     Vector<GPollFD, 32> pollFDs;
     while (!m_shouldStop) {
-        bool acquiredContext = g_main_context_acquire(mainContext);
-        ASSERT_UNUSED(acquiredContext, acquiredContext);
-
         gint maxPriority = 0;
         g_main_context_prepare(mainContext, &maxPriority);
 
@@ -120,12 +118,11 @@ void RunLoop::runGLibMainLoop(GMainContext* mainContext)
 
         g_main_context_check(mainContext, maxPriority, pollFDsData, numFDs);
         g_main_context_dispatch(mainContext);
-
-        g_main_context_release(mainContext);
     }
 
     notifyActivity(Activity::Exit);
     g_main_context_unref(mainContext);
+    g_main_context_pop_thread_default(mainContext);
 }
 
 void RunLoop::run()
```

nikolaszimmermann commented:
Thanks @mcatanzaro - I appreciate a thorough review, ... as you said: it's RunLoop! I'm currently in the process of also writing API tests for this. Will look at your last round of comments soon.

mcatanzaro commented:
OK, sorry, I see the problem. We do need to push/pop thread default for the _first_ nested level of the run loop. We do not need to do it for nested iterations. That's harmless, but overkill, because we already know it's the thread default (unless there is a bug in other code somewhere else, which is not our problem if so).

The way this function was originally written, we had two places where we push/pop. The first is needed, and the second is not needed:

```
void RunLoop::run()
{
    Ref runLoop = RunLoop::currentSingleton();
    GMainContext* mainContext = runLoop->m_mainContext.get();

    // The innermost main loop should always be there.
    ASSERT(!runLoop->m_mainLoops.isEmpty());

    GMainLoop* innermostLoop = runLoop->m_mainLoops[0].get();
    if (!g_main_loop_is_running(innermostLoop)) {
        g_main_context_push_thread_default(mainContext);
        g_main_loop_run(innermostLoop);
        g_main_context_pop_thread_default(mainContext);
        return;
    }

    // Create and run a nested loop if the innermost one was already running.
    GMainLoop* nestedMainLoop = g_main_loop_new(mainContext, FALSE);
    runLoop->m_mainLoops.append(adoptGRef(nestedMainLoop));

    g_main_context_push_thread_default(mainContext);
    g_main_loop_run(nestedMainLoop);
    g_main_context_pop_thread_default(mainContext);

    runLoop->m_mainLoops.removeLast();
}
```

So that's why we were talking about it not being needed. You've refactored it such that there is only one place we push/pop, and that place is needed. You could add checks to avoid it for the nested loops, but this would make the code more complicated, and simplicity is a virtue. So I think it's fine.

nikolaszimmermann commented:
I noticed a few problems while writing API tests, such as:
- RunLoop::cycle() doesn't send any notifications (but it should as on CF, it uses CFRunLoopWithMode, which supports observers)
- We don't properly support non-repeating observers

Will have all of that fixed in the next iteration, hopefully, while addressing all comments from @mcatanzaro.

nikolaszimmermann commented:
> I think we should add api tests for this.

Done, please check -- never added API tests so far :-)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b6325edc1a16bf2bb80d34b85061096bdea4713b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51678

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4e343b22c82f596898dcb50337d60cc5c1b0b65a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51678

carlosgcampos commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +85,73 @@ RunLoop::RunLoop()
+    g_main_context_dispatch(mainContext);
+}
 
+void RunLoop::runGLibMainLoop(GMainContext* mainContext)
```
Why do we pass the main context?

carlosgcampos commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -137,22 +161,84 @@ void RunLoop::wakeUp()
 
 RunLoop::CycleResult RunLoop::cycle(RunLoopMode)
 {
-    g_main_context_iteration(NULL, FALSE);
```
Wow, this was wrong always using the default main context.

nikolaszimmermann commented on Source/WTF/wtf/glib/ActivityObserver.h:
```diff
@@ -0,0 +1,63 @@
+
+// Activity observers (used to implement WebCore::RunLoopObserver)
+class ActivityObserver : public RefCounted<ActivityObserver> {
+    WTF_DEPRECATED_MAKE_FAST_ALLOCATED(ActivityObserver);
```
Fixed.

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +85,73 @@ RunLoop::RunLoop()
+    notifyActivity(Activity::BeforeWaiting);
+
+    if (auto* pollFunction = g_main_context_get_poll_func(mainContext)) {
+        auto result = (*pollFunction)(pollFDsData, numFDs, timeout);
```
This is embarrassing, good catch :/

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +85,73 @@ RunLoop::RunLoop()
+    gint maxPriority = 0;
+    g_main_context_prepare(mainContext, &maxPriority);
+
+    auto numFDs = g_main_context_query(mainContext, maxPriority, nullptr, nullptr, 0);
+    pollFDs.resize(numFDs);
+    auto* pollFDsData = pollFDs.mutableSpan().data();
+
+    gint timeout = 0;
+    g_main_context_query(mainContext, maxPriority, &timeout, pollFDsData, numFDs);
```
Well:

```
/host/home/nzimmermann/Software/GitRepositories/WebKit/Source/WTF/wtf/glib/RunLoopGLib.cpp:109:52: error: cannot convert ‘gint64*’ {aka ‘long int*’} to ‘gint*’ {aka ‘int*’}
  109 |     g_main_context_query(mainContext, maxPriority, &timeout, pollFDsData, numFDs);
      |                                                    ^~~~~~~~
      |                                                    |
      |                                                    gint64* {aka long int*}
```

```
/host/home/nzimmermann/Software/GitRepositories/WebKit/Source/WTF/wtf/glib/RunLoopGLib.cpp:109:52: error: cannot convert ‘gint64*’ {aka ‘long int*’} to ‘gint*’ {aka ‘int*’}
  109 |     g_main_context_query(mainContext, maxPriority, &timeout, pollFDsData, numFDs);
      |                                                    ^~~~~~~~
      |                                                    |
      |                                                    gint64* {aka long int*}
```

:-)

nikolaszimmermann replied:
Ok let's do it.

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/5bd4f8afc862a401dd621f00fadb240cccf25f9b)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124830 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44502 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35237 "Build is in progress. Recent messages:") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126707 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45205 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53071 "Build is in progress. Recent messages:") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127784 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/45205 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/35237 "Build is in progress. Recent messages:") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/45205 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/35237 "Build is in progress. Recent messages:") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116938 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/45205 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/35237 "Build is in progress. Recent messages:") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123350 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/53071 "Build is in progress. Recent messages:") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52090 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/35237 "Build is in progress. Recent messages:") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/35237 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48690 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51549 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57357 "Build is in progress. Recent messages:") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50941 "Build is in progress. Recent messages:") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54296 "Build is in progress. Recent messages:") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +85,74 @@ RunLoop::RunLoop()
 }
 
-void RunLoop::run()
+void RunLoop::runGLibMainLoopIteration(GMainContext* mainContext, MayBlock mayBlock)
```
Fixed.

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -137,22 +162,84 @@ void RunLoop::wakeUp()
+    wakeUp();
+}
+
+void RunLoop::unobserveActivity(const RefPtr<ActivityObserver>& observer)
```
`m_activityObservers.removeFirst(observer)` wouldn't work, if `observer` is not a `RefPtr<ActivityObserver>` - however I can convert all of them to Ref<ActivityObserver>. Going to try that.

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -137,22 +162,84 @@ void RunLoop::wakeUp()
+    m_eventObservers.add(observer);
+}
+
+void RunLoop::observeActivity(const RefPtr<ActivityObserver>& observer)
```
Converting all to Ref\<ActivityObserver\>.

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -137,22 +162,84 @@ void RunLoop::wakeUp()
+            observersToBeInvalidated.append(observer);
+    }
+
+    // Invalidation needs to happen _after_ dispatching all notifications.
+    {
+        Locker locker { m_activityObserversLock };
+        for (RefPtr observer : observersToBeInvalidated)
+            m_activityObservers.removeFirst(observer);
+    }
```
Yeah, stupid me. Was too late.

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -137,22 +162,84 @@ void RunLoop::wakeUp()
+
+    // Lock the activity observers, collect the ones to be notified.
+    {
+        Locker locker { m_activityObserversLock };
```
Sure! :-)

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -137,22 +162,84 @@ void RunLoop::wakeUp()
 
-    m_observers.forEach([event, name = String::fromUTF8(name)](auto& observer) {
+    m_eventObservers.forEach([event, name = String::fromUTF8(name)](auto& observer) {
         observer(event, name);
```
This is not a problem for the current use case, and we want to get rid of the event observers concept ideally.
Only `PageTimelineAgent` makes use of it:

```
    m_runLoopObserver = makeUnique<RunLoop::EventObserver>([this](RunLoop::Event event, const String& name) {
        if (!tracking() || m_environment.debugger()->isPaused())
            return;

        switch (event) {
        case RunLoop::Event::WillDispatch:
            pushCurrentRecord(TimelineRecordFactory::createRenderingFrameData(name), TimelineRecordType::RenderingFrame, false);
            break;
        case RunLoop::Event::DidDispatch:
            if (m_startedComposite)
                didComposite();
            didCompleteCurrentRecord(TimelineRecordType::RenderingFrame);
            break;
        }
    });
```

None of this mutates the set of event observers or adds/removes event observers -- therefore I concluded it's see to hold the lock indefinitely in `notifyEvent`.

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -137,22 +162,84 @@ void RunLoop::wakeUp()
+    {
+        Locker locker { m_activityObserversLock };
+
+        for (RefPtr observer : m_activityObservers) {
+            if (!observer->activities().contains(activity))
+                continue;
+            observersToBeNotified.append(observer);
+        }
```
Is this an improvment?

```
-        for (Ref observer : m_activityObservers) {
-            if (!observer->activities().contains(activity))
-                continue;
-            observersToBeNotified.append(observer);
-        }
+        observersToBeNotified = WTF::compactMap(m_activityObservers, [activity](const Ref<ActivityObserver>& observer) -> std::optional<Ref<ActivityObserver>> {
+            if (observer->activities().contains(activity))
+                return observer;
+            return std::nullopt;
+        });
```

To me this looks rather obfuscated.

I'd rather rewrite this without using continue:

```
        for (Ref observer : m_activityObservers) {
            if (observer->activities().contains(activity))
               observersToBeNotified.append(observer);
        }
```

This looks more concise to me.

nikolaszimmermann commented on Tools/TestWebKitAPI/glib/CMakeLists.txt:
```diff
@@ -163,6 +163,7 @@ ADD_WK2_TEST(TestWebsiteData ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestWeb
 ADD_WK2_TEST(TestDOMElement ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestDOMElement.cpp)
 ADD_WK2_TEST(TestGeolocationManager ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestGeolocationManager.cpp)
 ADD_WK2_TEST(TestInputMethodContext ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestInputMethodContext.cpp)
+ADD_WK2_TEST(TestRunLoopObserver ${TOOLS_DIR}/TestWebKitAPI/Utilities.cpp ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestRunLoopObserver.cpp)
```
I'd need to split between the `ActivityObserver` (glib-only, used to implement `RunLoopObserver`) and `RunLoopObserver` tests. This is actually a great idea, allows us to verify our understanding of RLO against Apple platforms.

nikolaszimmermann replied:
Done as well - all tests still pass on Gtk/WPE ports -- I'm eager to review the Mac results.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9da57b71b42e0b0ab4c1fe2ade1d361ae895d516. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51678

nikolaszimmermann commented:
Forgot to add TestRunLoopObserver to xcodeproj file -- fixed. Already observed some Gtk API test failures with the new TestRunLoopObserver - will investigate tomorrow.

nikolaszimmermann commented:
`RunLoopObserver` is very hard to test in a cross-platform way.

The Utilities::runFor(...) / ... functions in Tools/TestWebKitAPI exist in a mac variant (cocoa/Utilities.mm) that use NSRunLoop, not CFRunLoop -- no observers are fired if you use that one.

The cross-platform utilities in Utilities.cpp use `RunLoop::cycle()` in a while loop to drive the main loop - which is what I need for testing of the state transitions. However the macOS variants in cf/RunLoopCF.cpp they use a hardcoded **50ms** time window: the loop always spins for 50ms -- every 50ms (when called in a loop) state transitions will be fired (observer before/afterwaiting, entry/exit).

I've spent the day figuring out if there is a cross-platform solution, but it's not going to be trivial. I'll keep the tests for glib ports only at the moment.

carlosgcampos commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -86,48 +85,73 @@ RunLoop::RunLoop()
+    auto* pollFDsData = pollFDs.mutableSpan().data();
+
+    gint timeoutInMilliseconds = 0;
+    auto numFDs = g_main_context_query(m_mainContext.get(), maxPriority, &timeoutInMilliseconds, pollFDsData, pollFDs.size());
+    if (static_cast<size_t>(numFDs) > pollFDs.size()) {
+        pollFDs.grow(numFDs);
+        numFDs = g_main_context_query(m_mainContext.get(), maxPriority, &timeoutInMilliseconds, pollFDsData, pollFDs.size());
```
I still think we should iterate, another thread can attach a new source between the query calls

carlosgcampos replied:
https://github.com/WebKit/WebKit/pull/51678#discussion_r2405640323

nikolaszimmermann commented:
NOTE: There is a common set of crashes between gtk-wk2 / wpe-wk2, that cannot be coincidence.

```
Thread 1 (Thread 0x7f64242fd540 (LWP 2948206)):
#0  0x00007f642f02bda3 in WTF::fastFree(void*) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#1  0x00007f642f068db2 in WTF::RunLoop::performWork() () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#2  0x00007f642f168b6d in WTF::RunLoop::RunLoop()::{lambda(void*)#1}::_FUN(void*) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#3  0x00007f642f16b2ed in WTF::RunLoop::{lambda(_GSource*, int (*)(void*), void*)#1}::_FUN(_GSource*, int (*)(void*), void*) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#4  0x00007f642f4b449e in ??? () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#5  0x00007f642f4b4710 in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#6  0x00007f642f16c6a1 in WTF::RunLoop::runGLibMainLoopIteration(WTF::RunLoop::MayBlock) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#7  0x00007f642f16ccea in WTF::RunLoop::run() () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#8  0x00007f6434bee51e in int WebKit::AuxiliaryProcessMain<WebKit::WebProcessMainGtk>(int, char**) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#9  0x00007f642fa181ca in __libc_start_call_main (main=main@entry=0x561f345b77f0 <main>, argc=argc@entry=4, argv=argv@entry=0x7ffe0f7295c8) at ../sysdeps/nptl/libc_start_call_main.h:58
#10 0x00007f642fa1828b in __libc_start_main_impl (main=0x561f345b77f0 <main>, argc=4, argv=0x7ffe0f7295c8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffe0f7295b8) at ../csu/libc-start.c:360
#11 0x0000561f345b7725 in _start ()
```

and

```
Thread 1 (Thread 0x7fc8b6b0d540 (LWP 2893442)):
#0  0x00007fc8c182bda3 in WTF::fastFree(void*) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#1  0x00007fc8c68f703c in WTF::Vector<WebCore::HTTPHeaderMap::CommonHeader, 0ul, WTF::CrashOnOverflow, 6ul, WTF::FastMalloc>::shrink(unsigned long) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#2  0x00007fc8cac0e49a in WebCore::ResourceRequest::updateFromSoupMessageHeaders(_SoupMessageHeaders*) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#3  0x00007fc8c6ab92f5 in WebKit::WebSocketTask::WebSocketTask(WebKit::NetworkSocketChannel&, WebCore::ResourceRequest const&, _SoupSession*, _SoupMessage*, WTF::String const&)::{lambda(_SoupMessage*, WebKit::WebSocketTask*)#1}::_FUN(_SoupMessage*, WebKit::WebSocketTask*) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#4  0x00007fc8c1de12fa in g_closure_invoke () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#5  0x00007fc8c1e1090c in ??? () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#6  0x00007fc8c1e01591 in ??? () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#7  0x00007fc8c1e017c1 in g_signal_emit_valist () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#8  0x00007fc8c1e01883 in g_signal_emit () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#9  0x00007fc8bb4c48c3 in soup_message_starting (msg=<optimized out>) at ../../../../../jhbuild/checkout/libsoup/libsoup/soup-message.c:1259
#10 0x00007fc8bb4cd4d9 in soup_session_send_queue_item (session=<optimized out>, item=0x559351b27aa0, completion_cb=0x7fc8bb4d0310 <message_completed>) at ../../../../../jhbuild/checkout/libsoup/libsoup/soup-session.c:1408
#11 0x00007fc8bb4cdbb1 in soup_session_process_queue_item (session=0x559351a14b40, item=0x559351b27aa0, loop=1) at ../../../../../jhbuild/checkout/libsoup/libsoup/soup-session.c:1774
#12 0x00007fc8bb4ce020 in async_run_queue (session=<optimized out>) at ../../../../../jhbuild/checkout/libsoup/libsoup/soup-session.c:1852
#13 0x00007fc8bb4d049f in connect_async_complete (object=0x559351aba0b0, result=<optimized out>, user_data=<optimized out>) at ../../../../../jhbuild/checkout/libsoup/libsoup/soup-session.c:1653
#14 0x00007fc8c20cf79b in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#15 0x00007fc8c20d36f3 in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#16 0x00007fc8bb4b9f4f in connect_async_ready_cb (client=<optimized out>, result=0x559351ab8030, task=0x559351b105e0) at ../../../../../jhbuild/checkout/libsoup/libsoup/soup-connection.c:793
#17 0x00007fc8c20cf79b in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#18 0x00007fc8c20d36f3 in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#19 0x00007fc8c20c868c in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#20 0x00007fc8c20cf79b in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#21 0x00007fc8c20d36f3 in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#22 0x00007fc8c20ca1a0 in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#23 0x00007fc8c20baba1 in ??? () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#24 0x00007fc8c1ce049e in ??? () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#25 0x00007fc8c1ce0710 in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#26 0x00007fc8c196c6a1 in WTF::RunLoop::runGLibMainLoopIteration(WTF::RunLoop::MayBlock) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#27 0x00007fc8c196ccea in WTF::RunLoop::run() () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#28 0x00007fc8c6aa5bab in int WebKit::AuxiliaryProcessMain<WebKit::NetworkProcessMainSoup>(int, char**) () at /home/buildbot-worker/GTK-WK2-Tests-EWS/build/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#29 0x00007fc8c22181ca in __libc_start_call_main (main=main@entry=0x5593320617f0 <main>, argc=argc@entry=3, argv=argv@entry=0x7ffc9f647d28) at ../sysdeps/nptl/libc_start_call_main.h:58
#30 0x00007fc8c221828b in __libc_start_main_impl (main=0x5593320617f0 <main>, argc=3, argv=0x7ffc9f647d28, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffc9f647d18) at ../csu/libc-start.c:360
#31 0x0000559332061725 in _start ()
```

I managed to reproduce it locally - just a heads up that there's still a bug to fix.

nikolaszimmermann committed:
[GTK][WPE] Implement RunLoopObserver for glib ports
https://bugs.webkit.org/show_bug.cgi?id=300022

Reviewed by Carlos Garcia Campos.

Implement RunLoopObserver for GTK/WPE - ports using the GLib event loop.
This has been available on Apple platforms since a decade, and was used to
drive e.g. the rendering prior to GPUProcess. OpportunisticTaskScheduler
functionality (executing GC in 'good moments', not blocking rendering
etc.) depends on RunLoopObserver, which we failed to provide so far.

Add new API tests covering the ActiviyObserver/RunLoopObserver
implementation for the glib ports.

Tests: Tools/TestWebKitAPI/Tests/WTF/glib/ActivityObserver.cpp
       Tools/TestWebKitAPI/Tests/WebCore/glib/RunLoopObserver.cpp

* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/glib/ActivityObserver.h: Added.
(WTF::ActivityObserver::create):
(WTF::ActivityObserver::order const):
(WTF::ActivityObserver::activities const):
(WTF::ActivityObserver::notify const):
(WTF::ActivityObserver::ActivityObserver):
* Source/WTF/wtf/glib/RunLoopGLib.cpp:
(WTF::RunLoop::RunLoop):
(WTF::RunLoop::~RunLoop):
(WTF::RunLoop::runGLibMainLoopIteration):
(WTF::RunLoop::runGLibMainLoop):
(WTF::RunLoop::run):
(WTF::RunLoop::stop):
(WTF::RunLoop::cycle):
(WTF::RunLoop::observeEvent):
(WTF::RunLoop::observeActivity):
(WTF::RunLoop::unobserveActivity):
(WTF::RunLoop::notifyActivity):
(WTF::RunLoop::notifyEvent):
(WTF::RunLoop::observe): Deleted.
(WTF::RunLoop::notify): Deleted.
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp:
(WebCore::PageTimelineAgent::internalStart):
* Source/WebCore/inspector/agents/page/PageTimelineAgent.h:
* Source/WebCore/platform/RunLoopObserver.cpp:
(WebCore::RunLoopObserver::runLoopObserverFired):
* Source/WebCore/platform/RunLoopObserver.h:
(WebCore::RunLoopObserver::RunLoopObserver):
* Source/WebCore/platform/glib/RunLoopObserverGLib.cpp: Added.
(WebCore::RunLoopObserver::schedule):
(WebCore::RunLoopObserver::invalidate):
(WebCore::RunLoopObserver::isScheduled const):
* Tools/TestWebKitAPI/PlatformGTK.cmake:
* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WTF/glib/ActivityObserver.cpp: Added.
(TestWebKitAPI::dispatchChecker):
(TestWebKitAPI::runTestWhileRunLoopIsActive):
(TestWebKitAPI::TEST(WTF_ActivityObserver, Create)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, MatchingActivity)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, NonMatchingActivity)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, MultipleActivities)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, Ordering)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, SameOrder)):
* Tools/TestWebKitAPI/Tests/WebCore/glib/RunLoopObserver.cpp: Added.
(TestWebKitAPI::setUpTest):
(TestWebKitAPI::TEST(RunLoopObserver, Schedule)):
(TestWebKitAPI::TEST(RunLoopObserver, Invalidate)):
(TestWebKitAPI::TEST(RunLoopObserver, MultipleSchedule)):
(TestWebKitAPI::TEST(RunLoopObserver, MultipleInvalidate)):
(TestWebKitAPI::TEST(RunLoopObserver, Destruction)):
(TestWebKitAPI::TEST(RunLoopObserver, Repeating)):
(TestWebKitAPI::TEST(RunLoopObserver, OneShot)):
(TestWebKitAPI::TEST(RunLoopObserver, DefaultActivities)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityEntry)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityExit)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityBeforeWaiting)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityAfterWaiting)):
(TestWebKitAPI::TEST(RunLoopObserver, ActivityCombination)):
(TestWebKitAPI::TEST(RunLoopObserver, RemovesSelfDuringCallback)):
(TestWebKitAPI::TEST(RunLoopObserver, AddsNewObserverDuringCallback)):
(TestWebKitAPI::TEST(RunLoopObserver, AcrossMultipleIterations)):
(TestWebKitAPI::TEST(RunLoopObserver, WellKnownOrderValues)):
(TestWebKitAPI::TEST(RunLoopObserver, DifferentWellKnownOrderValues)):

Canonical link: https://commits.webkit.org/301378@main
